cci_cameroon/analysis/clustering_rumours/rumour_clustering_model_development.py

```diff
@@ -0,0 +1,1048 @@
+labelled_data.drop_duplicates(subset=["code", "comment"], inplace=True)
+print("After duplicate pairs removed: " + str(len(labelled_data)))
+
+# %%
+# Removing small count codes
+to_remove = list(
+    labelled_data.code.value_counts()[labelled_data.code.value_counts() < 10].index
+)
+labelled_data = labelled_data[~labelled_data.code.isin(to_remove)].copy()
+
+# %%
+# Dataset for modelling
+model_data = labelled_data.copy()
+# Create category ID column from the code field (join by _ into one string)
+model_data["category_id"] = model_data["code"].str.replace(" ", "_")
+id_to_category = dict(model_data[["category_id", "code"]].values)
+model_data = (
+    model_data.groupby("comment")["category_id"].apply(list).to_frame().reset_index()
+)
+
+# %%
+model_data = model_data.reset_index()
+
+# %%
+model_data.shape
+
+# %%
+english_df = data_df[data_df.language == "en"].copy()
+spanish_df = data_df[data_df.language == "es"].copy()
+data_df = data_df[~data_df.language.isin(["en", "es"])].copy()
+
+# %%
+english_df["comment"] = english_df.comment.apply(
+    translator.translate, src="en", dest="fr"
+).apply(getattr, args=("text",))
+spanish_df["comment"] = spanish_df.comment.apply(
+    translator.translate, src="es", dest="fr"
+).apply(getattr, args=("text",))
+
+# %%
+data = pd.concat([data_df, english_df, spanish_df], ignore_index=True)
+
+# %%
+data.drop("language", axis=1, inplace=True)
+
+# %%
+data.shape
+
+# %%
+data["category_id"] = data.first_code.factorize()[0]
+
+# %%
+# checking the distribution of the codes in the data
+data.groupby("first_code").comment.count().plot(kind="bar")
+
+# %%
+data = data.drop_duplicates(["comment"])
+data.reset_index(inplace=True)
+data.shape
+
+# %%
+# Converting the codes into integers to act as ground truth
+data["cluster"] = data.code.factorize()[0]
+
+# %%
+# using the french_semantic model for word embedding
+model = SentenceTransformer("Sahajtomar/french_semantic")
+# model = SentenceTransformer("distiluse-base-multilingual-cased-v1")
+
+# %%
+data[
+    data.category_id == 2
+].shape  # 127 cluster 1(0),137 cluster 2 (1), 136 cluster 3 (2)
+
+# %% [markdown]
+# ## Using a smaller sample of the data (~30 records) to ease inspection of results
+
+# %%
+# creating a small subset of the data made up of the different categories
+set1 = data[data.category_id == 0][["comment", "category_id"]][:10]
+set2 = data[data.category_id == 1][["comment", "category_id"]][:10]
+set3 = data[data.category_id == 2][["comment", "category_id"]][:10]
+to_use2 = (
+    pd.concat([set1, set2, set3]).drop_duplicates().reset_index().drop("index", axis=1)
+)
+# to_use2 = to_use.reset_index().drop("index",axis=1)
+
+# %%
+# create word embeddings using the transformer model
+sentence_embeddings2 = model.encode(to_use2.comment)
+
+# %%
+indp2 = faiss.index_factory(
+    sentence_embeddings2.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings2)  # to be used for inner product computation
```

Comment:
This line just keeps running for me. Unsure what the issue is and why it is so slow? Maybe we need someone else to test it their end too?